### PR TITLE
Store import nodes in SourceFile.imports instead of StringLiteral nodes

### LINKED ISSUE #20562: Debug Failure. False expression. in `getImportDeclaration ` during `getCodeActionForImport`
tsserver version: 2.6.1
hits: at least 351841 in the last week
stack:
```
Error: Debug Failure. False expression.
    at getImportDeclaration (/Applications/Visual Studio Code.app/Contents/Resources/app/extensions/node_modules/typescript/lib/tsserver.js:75926:30)
    at /Applications/Visual Studio Code.app/Contents/Resources/app/extensions/node_modules/typescript/lib/tsserver.js:75913:98
    at Object.mapDefined (/Applications/Visual Studio Code.app/Contents/Resources/app/extensions/node_modules/typescript/lib/tsserver.js:1610:30)
    at getImportDeclarations (/Applications/Visual Studio Code.app/Contents/Resources/app/extensions/node_modules/typescript/lib/tsserver.js:75912:72)
    at Object.getCodeActionForImport (/Applications/Visual Studio Code.app/Contents/Resources/app/extensions/node_modules/typescript/lib/tsserver.js:75882:32)
    at getCompletionEntryCodeActions (/Applications/Visual Studio Code.app/Contents/Resources/app/extensions/node_modules/typescript/lib/tsserver.js:65309:31)
    at Object.getCompletionEntryDetails (/Applications/Visual Studio Code.app/Contents/Resources/app/extensions/node_modules/typescript/lib/tsserver.js:65283:35)
    at Object.getCompletionEntryDetails (/Applications/Visual Studio Code.app/Contents/Resources/app/extensions/node_modules/typescript/lib/tsserver.js:79782:35)
    at /Applications/Visual Studio Code.app/Contents/Resources/app/extensions/node_modules/typescript/lib/tsserver.js:86221:64
    at Object.mapDefined (/Applications/Visual Studio Code.app/Contents/Resources/app/extensions/node_modules/typescript/lib/tsserver.js:1610:30)
    at IOSession.Session.getCompletionEntryDetails (/Applications/Visual Studio Code.app/Contents/Resources/app/extensions/node_modules/typescript/lib/tsserver.js:86219:27)
    at Session.handlers.ts.createMapFromTemplate._a.(anonymous function) (/Applications/Visual Studio Code.app/Contents/Resources/app/extensions/node_modules/typescript/lib/tsserver.js:85317:61)
    at /Applications/Visual Studio Code.app/Contents/Resources/app/extensions/node_modules/typescript/lib/tsserver.js:86669:88
    at IOSession.Session.executeWithRequestId (/Applications/Visual Studio Code.app/Contents/Resources/app/extensions/node_modules/typescript/lib/tsserver.js:86660:28)
    at IOSession.Session.executeCommand (/Applications/Visual Studio Code.app/Contents/Resources/app/extensions/node_modules/typescript/lib/tsserver.js:86669:33)
    at IOSession.Session.onMessage (/Applications/Visual Studio Code.app/Contents/Resources/app/extensions/node_modules/typescript/lib/tsserver.js:86689:35)
    at Interface.<anonymous> (/Applications/Visual Studio Code.app/Contents/Resources/app/extensions/node_modules/typescript/lib/tsserver.js:87881:27)
    at emitOne (events.js:96:13)
    at Interface.emit (events.js:191:7)
    at Interface._onLine (readline.js:241:10)
    at Interface._normalWrite (readline.js:384:12)
    at Socket.ondata (readline.js:101:10)
    at emitOne (events.js:96:13)
    at Socket.emit (events.js:191:7)
    at readableAddChunk (_stream_readable.js:178:18)
    at Socket.Readable.push (_stream_readable.js:136:10)
    at Pipe.onread (net.js:560:20)
```

### PR DESCRIPTION
**Pull request by andy-ms**
Fixes #20562 
Previously `SourceFile.imports` was a list of `StringLiteral`s and we had to climb parents to get to the importer node, which is error-prone as `node.parent` is just a `Node`. Now we store the import nodes directly and climb *down* if we need the `StringLiteral`.

### CODE REVIEW

**Comment by andy-ms** [comment]
@sheetalkamat @DanielRosenwasser @aozgaa Could someone review this? Underlying issue is marked 2.8.1.

**Comment by sheetalkamat** [comment]
I prefer that we just handle the scenario of getting import declaration from string literal correctly rather than changing the storage from module specifiers to import declarations

**Comment by RyanCavanaugh** [comment]
No tests?

**Comment by andy-ms** [comment]
This was a telemetry issue.. I haven't been able to figure out a case that actually fails at runtime.

**Comment by andy-ms** [comment]
@sheetalkamat Please re-review

File: src/compiler/types.ts [hunk 1]
```diff
@@ -3146,6 +3146,18 @@ namespace ts {
     /* @internal */
     export type AnyImportSyntax = ImportDeclaration | ImportEqualsDeclaration;
 
+    /* @internal */
+    export type AnyImportOrReExport = AnyImportSyntax | ExportDeclaration;
+
+    /* @internal */
+    export type AnyValidImportOrReExport =
+        | (ImportDeclaration | ExportDeclaration) & { moduleSpecifier: StringLiteral }
```
Comment (sheetalkamat) [inline src/compiler/types.ts]: This seems to be incorrect from https://github.com/Microsoft/TypeScript/pull/22495/files#diff-08a3cc4f1f9a51dbb468c2810f5229d3R1607 where moduleSpecifier is not present ?

File: src/compiler/utilities.ts [hunk 2]
```diff
@@ -1650,21 +1654,29 @@ namespace ts {
...
-                return reference.expression;
-            }
-        }
-        if (node.kind === SyntaxKind.ExportDeclaration) {
-            return (<ExportDeclaration>node).moduleSpecifier;
+    export function importFromModuleSpecifier(node: StringLiteralLike): AnyValidImportOrReExport {
+        switch (node.parent.kind) {
+            case SyntaxKind.ImportDeclaration:
+            case SyntaxKind.ExportDeclaration:
+                return node.parent as AnyValidImportOrReExport;
```
Comment (sheetalkamat) [inline src/compiler/utilities.ts]: And this is where we might crash as moduleSpecifier is absent on import declaration at: https://github.com/Microsoft/TypeScript/pull/22495/files#diff-08a3cc4f1f9a51dbb468c2810f5229d3R1607 ?

File: src/compiler/program.ts [hunk 3]
```diff
@@ -1621,66 +1621,55 @@ namespace ts {
...
                 }
             }
 
             function collectDynamicImportOrRequireCalls(node: Node): void {
                 if (isRequireCall(node, /*checkArgumentIsStringLiteral*/ true)) {
-                    (imports || (imports = [])).push(<StringLiteral>(<CallExpression>node).arguments[0]);
+                    imports = append(imports, node.arguments[0]);
                 }
                 // we have to check the argument list has length of 1. We will still have to process these even though we have parsing error.
                 else if (isImportCall(node) && node.arguments.length === 1 && node.arguments[0].kind === SyntaxKind.StringLiteral) {
```
Comment (sheetalkamat) [inline src/compiler/program.ts]: Do we need to check instead ```isStringLiteralLike(node.arguments[0])``` instead of checking just string literal ?

Comment (andy-ms) [reply src/compiler/program.ts]: `isRequireCall` checks for `StringLiteralLike`, updated the parameter name

File: src/compiler/types.ts (hunk 1 above)
Comment (andy-ms) [reply src/compiler/types.ts]: It should be, added it there.

File: src/compiler/program.ts (hunk 3 above)
Comment (sheetalkamat) [reply src/compiler/program.ts]: No i meant 
``` typescript
// we have to check the argument list has length of 1. We will still have to process these even though we have parsing error.
                 else if (isImportCall(node) && node.arguments.length === 1 && node.arguments[0].kind === SyntaxKind.StringLiteral) {

```

Comment (andy-ms) [reply src/compiler/program.ts]: Oh, sure. Now we can get better typings for:
```ts
const foo = await import(`foo`);
```

**Pull request merged**